AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +621,834 @@ class AzureIoTHub {
+            }
+            _isConnecting = false;
+            // TODO: What should we pass to _onConnectCb?
+            _onConnectCb(err);
+        }
+
+        function _onDisconnect() {
+            _log("Disconnected!");
+            local reason = _isDisconnecting ? 0 : AZURE_IOT_CLIENT_ERROR_GENERAL;
+            _cleanup();
+            _onDisconnectCb && _onDisconnectCb(reason);
+        }
+
+        function _onMessage(msg) {
+            local message = null;
+            local topic = null;
+            try {
+                message = msg["message"];
+                topic = msg["topic"];
+                _log(format("_onMessage: topic=%s | body=%s", topic, message));
+            } catch (e) {
+                _logError("Could not read message: " + e);
+                return;
             }
-        }
-
-        // Helper Methods
-        // ------------------------------------------------------------------------------------
 
-        function _isOpen(amqpObj) {
-            return (amqpObj != null && amqpObj.isopen());
+            // Cloud-to-device message received
+            if (topic.find(_topics.msgRecv) != null) {
+                _handleCloudToDevMsg(message, topic);
+            // Desired properties were updated
+            } else if (topic.find(_topics.twinNotif) != null) {
+                _handleDesPropsMsg(message, topic);
+            // Twin request result received
+            } else if (topic.find(_topics.twinRecv) != null) {
+                _handleTwinResponse(message, topic);
+            // Direct method called
+            } else if (topic.find(_topics.dMethodNotif) != null) {
+                _handleDirMethodMsg(message, topic);
+            }
         }
 
-        function _connectionCallback(error) {
-            _connecting = false;
-            if (_handlers.onConnected) {
-                imp.wakeup(0, function() {
-                    _handlers.onConnected(error);
-                }.bindenv(this));
+        function _handleCloudToDevMsg(message, topic) {
+            local props = {};
+            local splittedTopic = split(topic, "/");
+            if (splittedTopic[splittedTopic.len() - 1] != "devicebound") {
+                // We have properties sent with message
+                props = http.urldecode(splittedTopic[splittedTopic.len() - 1]);
+            }
+            _onMessageCb(AzureIoTHub.Message(message, props));
+        }
+
+        function _handleDesPropsMsg(message, topic) {
+            local version = null;
+            local parsedMsg = null;
+            try {
+                version = split(topic, "=")[1];
+                parsedMsg = http.jsondecode(message);
+            } catch (e) {
+                _logError("Exception at parsing the message: " + e);
+                _logMsg(message, topic);
+                return;
+            }
+            _onTwinReqCb(version, parsedMsg);
+        }
+
+        function _handleTwinResponse(message, topic) {
```

Comment:
This method should be refactored to a) make it shorter and b) more readable. For example there are two clear subroutines that can be extracted.

Reply:
Done